cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
+                task.add_broadcast(
+                    output_arr.base, axes=tuple(range(1, len(out_shape)))
+                )
+        else:
+            task.add_scalar_arg(self.ndim, ty.int64)
+            task.add_scalar_arg(self.ndim, ty.int64)
+            task.add_scalar_arg(start_index, ty.int64)
+            for index_arr in arrays:
+                task.add_input(index_arr)
+                task.add_alignment(output_arr.base, index_arr)
+        task.execute()
+
+        return output_arr
+
+    def _copy_store(self, store):
+        store_to_copy = DeferredArray(
+            self.runtime,
+            base=store,
+            dtype=self.dtype,
+        )
+        store_copy = self.runtime.create_empty_thunk(
+            store_to_copy.shape,
+            self.dtype,
+            inputs=[store_to_copy],
+        )
+        store_copy.copy(store_to_copy, deep=True)
+        return store_copy, store_copy.base
+
+    def _create_indexing_array(self, key, is_set=False):
+        store = self.base
+        rhs = self
+        # the index where the first index_array is passed to the [] operator
+        start_index = -1
         if isinstance(key, tuple):
+            key = self._unpack_ellipsis(key, self.ndim)
+            shift = 0
+            last_index = self.ndim
+            # in case when index arrays are passed in the scaterred way,
+            # we need to transpose original array so all index arrays
+            # are close to each other
+            transpose_needed = False
+            transpose_indices = tuple()
+            # since we can't call Copy operation on transformed Store, after
+            # the transformation, we need to return a copy
             tuple_of_arrays = ()
-            for k in key:
-                if not isinstance(k, NumPyThunk):
-                    raise NotImplementedError(
-                        "need support for mixed advanced indexing"
+            index_map = []
+
+            # First, we need to check if transpose is needed
+            for dim, k in enumerate(key):
+                if np.isscalar(k) or isinstance(k, NumPyThunk):
+                    if start_index == -1:
+                        start_index = dim
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
```

Comment:
I think this is simpler to write as:

```
for count, i in enumerate(transpose_indices):
  index_map[i] = count
```

Reply:
Actually, is it simpler to just transpose the key here, to match the store?
```
key = tuple(key[i] for i in transpose_indices)
```
Then you don't need `index_map` I think.

Reply:
Sure, thank you